nd4j/nd4j-serde/nd4j-arrow/src/main/java/org/nd4j/arrow/ByteDecoArrowSerde.java

```diff
@@ -0,0 +1,385 @@
+        }
+        else if(dataType.equals(arrow.date64())) {
+            throw new IllegalArgumentException("Unable to convert type " + dataType.name());
+        }
+        else if(dataType.equals(arrow.day_time_interval())) {
+            throw new IllegalArgumentException("Unable to convert type " + dataType.name());
+
+        }
+        else if(dataType.equals(arrow.large_utf8())) {
+            return org.nd4j.linalg.api.buffer.DataType.UTF8;
+        }
+        else if(dataType.equals(arrow.utf8())) {
+            return org.nd4j.linalg.api.buffer.DataType.UTF8;
+        }
+        else if(dataType.equals(arrow.binary())) {
+            return org.nd4j.linalg.api.buffer.DataType.BYTE;
+        }
+        else {
+            throw new IllegalArgumentException("Unable to convert type " + dataType.name());
+        }
+    }
+
+    /**
+     *
+     * @param arrowBuffer
+     * @param dataType
+     * @return
+     */
+    public static DataBuffer fromArrowBuffer(ArrowBuffer arrowBuffer,DataType dataType) {
+        org.nd4j.linalg.api.buffer.DataType dataType1 = dataBufferTypeTypeForArrow(dataType);
+        if(dataType1 != org.nd4j.linalg.api.buffer.DataType.UTF8) {
+            BytePointer bytePointer = arrowBuffer.data().capacity(arrowBuffer.size() * dataType1.width());
+            return Nd4j.createBuffer(bytePointer,arrowBuffer.size(),dataBufferTypeTypeForArrow(dataType));
+
+        }
+        else {
+            BytePointer bytePointer = arrowBuffer.data();
+            return Nd4j.createBuffer(bytePointer,arrowBuffer.size(),dataBufferTypeTypeForArrow(dataType));
+
+        }
+    }
+
+    /**
+     * Create a {@link Pair}
+     * of {@link ArrowBuffer} and {@link org.nd4j.linalg.api.buffer.DataType}
+     * based on the input {@link DataBuffer}
+     * @param dataBuffer the input data buffer
+     * @return the pair
+     */
+    public static Pair<ArrowBuffer,DataType> fromNd4jBuffer(DataBuffer dataBuffer) {
+        BytePointer bytePointer = new BytePointer(dataBuffer.pointer());
+        ArrowBuffer arrowBuffer = new ArrowBuffer(bytePointer,dataBuffer.length());
+        return Pair.of(arrowBuffer,arrowDataTypeForNd4j(dataBuffer.dataType()));
+    }
+
+
+    /**
+     * Creates an {@link INDArray} from an arrow {@link Array}
+     * @param array the input {@link Array}
+     * @return the equivalent {@link INDArray} zero copied
+     */
+    public static INDArray ndarrayFromArrowArray(FlatArray array) {
+        if(array instanceof PrimitiveArray) {
+            PrimitiveArray primitiveArray = (PrimitiveArray) array;
+            ArrowBuffer arrowBuffer = primitiveArray.values();
+            DataBuffer nd4jBuffer = fromArrowBuffer(arrowBuffer,array.data().type());
+            return Nd4j.create(nd4jBuffer,1,nd4jBuffer.length());
+        }
+        else {
+            StringArray stringArray = (StringArray) array;
+            ArrowBuffer arrowBuffer = stringArray.value_data().capacity(array.capacity()).limit(array.limit());
```

Comment:
And here too, `StringArray.limit()/capacity()` most likely isn't what you want.

Reply:
@saudet ok I got rid of this.